### PR TITLE
Consolidate Beacon config

### PR DESCRIPTION
**Pull request by mpetrunic**
Implemented:
- separated config interfaces per module
- separated default values per module
- introduced configuration description
- generation and validaton of toml config using configuration description

TODO:
- generate cli options from configuration description
- start rpc depending on config - host, port, transport, apis (I would open a separate issue for that as it's not so simple would be useful to implement rest api before that)

### CODE REVIEW

**Comment by mpetrunic** [comment]
@GregTheGreek @wemeetagain Just want a quick look before I implement cli part

**Comment by GregTheGreek** [comment]
This looks like it will conflict with #253 pretty hard

**Comment by mpetrunic** [comment]
> This looks like it will conflict with #253 pretty hard

Yup, I'm vaiting for it to be merged and I will adjust everything

**Comment by wemeetagain** [comment]
I think you should consider aligning IConfigurationField with the yargs Option.
We can remove a lot of the validation, or push it to the configuration/cli parser library if we use yargs.

**Comment by mpetrunic** [comment]
> I think you should consider aligning IConfigurationField with the yargs Option.
> We can remove a lot of the validation, or push it to the configuration/cli parser library if we use yargs.

@wemeetagain I started with aligning with yargs options but it introduced a few issues:
- those validation is yargs specific so I can't validate toml config
- it has a lot of fields that can be implicitly deducted like group -> parent name, alias -> field name

I can easily generate cli options and validations from current props and thus use the same validation for toml config. 

@GregTheGreek @wemeetagain Should I switch to yargs in this PR or in separate?

**Comment by wemeetagain** [comment]
I think a separate PR would be better. It will probably be good sized code diff.
You should be able to read a toml configuration into yargs, the idea was then you won't have to maintain separate validation code. Thinking using [this](https://yargs.js.org/docs/#api-configobject) with the second parameter doing the raw toml file parse.
The more explicit style lets us be more flexible. Even if some of the fields are filled in dynamically (eg: group name), things like alias might still be useful to explicitly define.

**Comment by mpetrunic** [comment]
> You should be able to read a toml configuration into yargs, the idea was then you won't have to maintain separate validation code. Thinking using [this](https://yargs.js.org/docs/#api-configobject) with the second parameter doing the raw toml file parse.

Hmm but you would want different option names in cli. Example db name, this would imply that cli accepts --name while it should be --beaconDB. 

When you load toml into yargs who will have precedence, currently cli can override some toml conf.

I don't think we will achieve anything with adjusting everything to yargs, I think this way gives a lot more flexibility in terms of custom types and validation (example verify that it is ethereum address)

**Comment by wemeetagain** [comment]
I think the toml parsing step would transform the input in a uniform way. eg: every option under the db header would have "db" appended to the option name. The cannonical name of each modules options would be prefixed with the name of the module.

I'm thinking the cli should definitely override toml settings, like youre suggesting. This is the current behavior of the yargs config method.

Custom validation can be done by extending the yargs Option interface with a validation key (like you have in this pr) and using a yargs middleware to apply the extra validations.

What we would achieve by using the yargs machinery is a consistent unified approach to handling all user configuration and validation. Its the same ideal as this pr, just letting the library handle as much as it can (which is almost everything).

**Comment by mpetrunic** [comment]
I very much dislike yargs validation, as it's very hacky to extend is using middleware. There is a lot of edge cases, example arrays in yargs are represented as string with comas, while toml supports arrays out of the box so I'm not really sure what would happen if we are to pass array instead of string to yargs,

I would rather use something like https://github.com/hapijs/joi for validation as it's much cleaner and sane way to validate rather than defining validation as bunch of fields.

That said I will use current config as it doens't really include much validation to generate cli using commander and open issue for switching to yargs and validation so we can discuss there and implement in another PR

**Comment by GregTheGreek** [comment]
> Should I switch to yargs in this PR or in separate?

Seperate for sure!

> I would rather use something like https://github.com/hapijs/joi for validation as it's much cleaner and sane way to validate rather than defining validation as bunch of fields.

I think I would be ok with this 👍

**Comment by wemeetagain** [comment]
> I very much dislike yargs validation, as it's very hacky to extend is using middleware. 

I agree that the yargs types/validation isn't very full-featured for what we want. What I liked about it is the integration with cli documentation.

> There is a lot of edge cases, example arrays in yargs are represented as string with comas, while toml supports arrays out of the box so I'm not really sure what would happen if we are to pass array instead of string to yargs,

The yargs config method expects input parsed into a structural form. So in that case, an option that should be an array would expect an Array to be passed in.

I think my main concerns / priorities with all this is that:
- we don't want to be building any special-cased validation for one 'type' of user input
  - we want cli/config/env-vars/etc to all be run through the same validation/etc steps
- we don't want to be building special code per option for each 'type' of user input
  - we want one schema that can translate to how options are presented in cli/config/env-vars/etc
- we want to (practically, not militantly) minimize the amount of code we're maintaining, especially for non-ethereum tasks
  - if there's a great library that can handle what we're doing, lets use it

Beyond that, not really worried about which direction we go.

> I would rather use something like https://github.com/hapijs/joi for validation as it's much cleaner and sane way to validate rather than defining validation as bunch of fields.

Agreed. It looks great!

> That said I will use current config as it doens't really include much validation to generate cli using commander and open issue for switching to yargs and validation so we can discuss there and implement in another PR

:+1:

**Comment by mpetrunic** [comment]
I kinda like configuration now :)

File: src/eth1/impl/ethers.ts [hunk 1]
```diff
@@ -44,7 +44,14 @@ export class EthersEth1Notifier extends EventEmitter implements IEth1Notifier {
     super();
     this.logger = logger;
     this.opts = opts;
-    this.provider = opts.provider;
+    if(this.opts.providerInstance) {
+      this.provider = this.opts.providerInstance  ;
```
Comment (wemeetagain) [inline src/eth1/impl/ethers.ts]: extra space

File: src/cli/util.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    } else {
+      processField(field, options, config);
+    }
+  });
+  return config;
+}
+
+function processModule(options: { [p: string]: string }, field, config) {
+  const childConfig = optionsToConfig(options, field as IConfigurationModule);
+  if (Object.keys(childConfig).length > 0) {
```
Comment (wemeetagain) [inline src/cli/util.ts:49]: Why is this check needed here?

File: src/cli/util.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+function processField(field, options: { [p: string]: string }, config) {
+  field = field as IConfigurationField<unknown>;
+  if (field.cli && options[field.cli.flag]) {
+    let value: any = options[field.cli.flag];
+    if (field.process && typeof value === 'string') {
+      value = field.process(value);
+    }
+    if (!field.validation || field.validation(value)) {
```
Comment (wemeetagain) [inline src/cli/util.ts]: I think we're wanting to throw an error on invalid user input

File: src/cli/commands/deposit.ts [hunk 4]
```diff
@@ -89,7 +89,8 @@ export class DepositCommand implements CliCommand {
             `Successfully deposited ${options.value} ETH from ${wallet.address} 
             to deposit contract. Tx hash: ${hash}`
           );
-        } catch (e) {
+        }
```
Comment (wemeetagain) [inline src/cli/commands/deposit.ts]: spacing

**Review by wemeetagain — CHANGES_REQUESTED** [review]
Wow, significantly better than what we had before!

File: src/cli/util.ts (hunk 2 above)
Comment (mpetrunic) [reply src/cli/util.ts:49]: If you pass empty object to toml library it produces weird invalid toml

File: src/cli/util.ts (hunk 3 above)
Comment (mpetrunic) [reply src/cli/util.ts]: throw or just log warning which prop is invalid?
 I mean default prop will kick in so it doesn't necessarily have to stop node from starting

File: src/cli/commands/beacon.ts [hunk 5]
```diff
@@ -4,137 +4,91 @@
...
-  eth1RpcUrl: string;
-  rpc: string;
-  configFile: string;
+  configFile?: string;
   loggingLevel?: string;
-  validator?: {
-    key: string;
-    db?: string;
-  };
+  [key: string]: string;
```
Comment (GregTheGreek) [inline src/cli/commands/beacon.ts:20]: what is this?

**Review by GregTheGreek — CHANGES_REQUESTED** [review]

File: src/cli/commands/beacon.ts (hunk 5 above)
Comment (mpetrunic) [reply src/cli/commands/beacon.ts:20]: it allows any key value property so we dont have to define generated options again here. it gets validate later so everything unsupported gets remove

Comment (GregTheGreek) [reply src/cli/commands/beacon.ts:20]: why not just use a map?

Comment (mpetrunic) [reply src/cli/commands/beacon.ts:20]: I wan't to have configFile and loggingLevel props typed

**Review by wemeetagain — APPROVED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Pull request merged**